tensorizer/serialization.py

```diff
@@ -635,6 +1090,8 @@ class TensorDeserializer(
         verify_hash: If True, the hashes of each tensor will be verified
             against the hashes stored in the metadata. A `HashMismatchError`
             will be raised if any of the hashes do not match.
+        encryption: A `DecryptionParams` object holding a password or key
```

Comment:
>Can we have a passphrase option here, and check if it's a string? If it's a string, we should be able to create this internally.

This sounds like an extremely bad idea while only the "fast" (should really be named "insecure") key derivation function is provided. Users should not be tricked into using it without explicitly choosing to ignore the security implications. Unless this is supposed to be just for providing a hex representation of a proper random key?

Reply:
> Can we have a passphrase option here, and check if it's a string? If it's a string, we should be able to create this internally. I'm not a fan of OOP like this that requires the instantiation of a parameters object.

Responding to the original suggestion: the best way to use this is with `DecryptionParams.from_key()` (with `EncryptionParams.random()`), since it will give the best security and the best performance (from not requiring re-derivation of the key at deserialization time). I'm not a fan of providing a quick-route to the passphrase option, since it is always worse either in security or performance (even with a proper KDF) and providing a shortcut for it is implying that "this is easiest because this is the usual way you should be doing it anyway." I would rather that be a conscious decision on the part of the user, done by picking a `DecryptionParams` factory function.

Its usage is symmetric with `EncryptionParams` as well, which must always be chosen by the user, and (in my opinion) the consistency between the two improves code legibility compared to the types being different, as in:
```py
# Serialization
encryption_params = EncryptionParams.from_passphrase_fast(passphrase)

serializer = TensorSerializer("model.tensors", encryption=encryption_params)
serializer.write_module(...)
serializer.close()

# Deserialization
decryption_params = DecryptionParams.from_passphrase(passphrase)

deserializer = TensorDeserializer("model.tensors", encryption=decryption_params)
deserializer.load_into_module(...)
deserializer.close()
```
Versus:
```py
# Serialization
encryption_params = EncryptionParams.from_passphrase_fast(passphrase)

serializer = TensorSerializer("model.tensors", encryption=encryption_params)
serializer.write_module(...)
serializer.close()

# Deserialization
deserializer = TensorDeserializer("model.tensors", encryption=passphrase)
deserializer.load_into_module(...)
deserializer.close()
```

Reply:
> I was referring to to the suggestion of making it accept a string instead of a `DecryptionParams` object. If it is enforced that this string is e.g. 32 bytes in hex or base64, that would be fine. If it would transparently call `from_passphrase_fast`, it would be a trap for users.

The key derivation method is chosen during serialization time, so using `EncryptionParams.from_passphrase_fast()` in a `TensorSerializer` will essentially encode in the file that "this tensor is eligible to be decrypted by using `from_passphrase_fast()`." At deserialization time, the only options are `DecryptionParams.from_key()`, meaning "skip using any KDF," and `DecryptionParams.from_passphrase(),` meaning "use an unspecified KDF based on the file being read." When you use `DecryptionParams.from_passphrase()`, it will begin to read the tensor, and then turn the passphrase into a key using the same algorithm it was originally encrypted with once that information is found. Since the suggestion was for the deserializer, it wouldn't be able to choose `from_passphrase_fast` for you unless you had already used it during serialization.

If the original suggestion was intended to apply to the `TensorSerializer` as well then that *would* require choosing a default KDF for encryption, which I would definitely consider a no-no.

Reply:
Updates (@wbrown/@finetunej):

Argon2id (via [`libsodium`'s `pwhash` API](https://libsodium.gitbook.io/doc/password_hashing/default_phf)) is now available as a key derivation function, `EncryptionParams.from_string(source, opslimit, memlimit, salt)`. `DecryptionParams.from_passphrase(passphrase)` was renamed to `DecryptionParams.from_string(source)` to match.

Documentation for these are available in `docs/encryption.md` with the rest of the documentation for encryption, as well as in docstrings.

`EncryptionParams.from_passphrase_fast()` was completely removed. Argon2id is configurable enough that its fastest mode has comparable speed and down to 8 KiB of RAM usage, if needed. An exact key can always be supplied via `EncryptionParams(key=...)` derived any way one would like for the remaining extremely specific use cases.

Argon2/`pwhash` parameters used during serialization are encoded in a `PWHashKeyDerivationChunk` (`derivation_method=1`) that stores `opslimit`, `memlimit`, `salt`, and `alg` (currently always `crypto_pwhash_ALG_ARGON2ID13`) parameters from the `pwhash` API, so that all of those parameters can be automatically inferred during deserialization.